Update cat_facts example for Crux 0.17, and update APIs for crux_kv, crux_platform and crux_time

StuartHarris opened the pull request:
_note: these are breaking changes_

## `crux_kv`, `crux_platform`, `crux_time`

- [x] Add `#[derive(Facet)]` where missing
- [x] Move all protocol types to their own module, to make it clear that these are the only types that need to implement `serde` and `facet` traits.

## `cat_facts` example

- [x] removes shared types lib
- [x] adds `CoreFFI`
    - [x] uses `crux_core::bridge::Bridge`
    - [x]  introduces new features for `uniffi` and `wasm_bindgen`, which means that `web-yew` and `cli` shells no longer include a bridge
- [x] add codegen binary, for facet based typegen of Swift, Kotlin, TypeScript, and bindgen for Android
- [x] use `resolver = "3"`
- [x] Use `#[facet(opaque)]` to allow us to use types that don't implement the `Facet` trait, in event variants that don't leave the shell (depends on the newly released `facet_generate` v0.12.1)
- [x] update `iOS` app, uses `uniffi` feature
- [x] update `Android` app (including deps, moving from gradle to kts), uses `uniffi` feature
- [x] update `web-nextjs` (including wasm package in generated folder), uses `wasm_bindgen` feature
- [x] update `web-yew`, doesn't use `uniffi` or `wasm_bindgen` features, so no bridge
- [x] update cli, doesn't use `uniffi` or `wasm_bindgen` features, so no bridge
- [x] adds a `Justfile` for each shell, so that you can just `cd` into the relevant directory and run `just dev` to get going.

charypar commented on crux_kv/src/lib.rs:
```diff
@@ -87,26 +85,28 @@ impl std::fmt::Debug for KeyValueOperation {
     /// Response to a `KeyValueOperation::Exists`,
     /// returning whether the key is present in the store
-    Exists { is_present: bool },
+    Exists(bool),
```
I'm reasonably sure we're going back to what we once had, which we changed to be more explicit about what the carried values are, especially for the `previous` case. 

If these are purely a matter of taste, I think I'd leave them alone, especially since this is probably quite a heavily used type, so the breaking change is high impact.

charypar commented on crux_kv/src/value.rs:
```diff
@@ -1,37 +0,0 @@
-#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
-pub enum Value {
-    None,
-    Bytes(#[serde(with = "serde_bytes")] Vec<u8>),
```
This was important for two reasons: 

1. signalling not found
2. serializing bytes better with formats that have special handling for them.

charypar commented:
Yea, we do need to own the protocol types, and they also need to serialize well into various formats (esp. JSON which will probably be common). That's also what then makes them kinda nasty to use in normal Rust code, they are really more of an FFI format...

StuartHarris commented:
The confusion was due to inward-facing types needing to implement `Facet` in order to be part of the `Event` enum. I've since updated `facet_generate` to support `#[facet(opaque)]` correctly, so that now fields in structs and in struct or tuple variants can reference types that don't need to implement `Facet`. This means we can use the capabilities correctly (although I moved all the protocol types to a protocol module, and implemented `Facet` on them where missing).

charypar reviewed (approved):
Amazing, thanks for doing all that! ❤️

StuartHarris commented on crux_platform/src/lib.rs:
I would love to. It does nothing lol

StuartHarris commented on examples/cat_facts/shared/src/app.rs:
```diff
@@ -54,34 +55,43 @@ impl Default for CatImage {
+    // we can skip typegen using `#[facet(skip)]`
+    // we can use types that do not implement `Facet`, by marking them with `#[facet(opaque)]`
     #[serde(skip)]
+    #[facet(skip)]
```
Huh. I hadn't thought of that but a good idea if we can't read them directly. I'll have another look at this.
